signerv2/kms_signer.go

```diff
@@ -0,0 +1,103 @@
+package signerv2
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"errors"
+	"math/big"
+
+	eigenkms "github.com/Layr-Labs/eigensdk-go/aws/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/secp256k1"
+)
+
+var secp256k1N = crypto.S256().Params().N
+var secp256k1HalfN = new(big.Int).Div(secp256k1N, big.NewInt(2))
+
+func NewKMSSigner(ctx context.Context, svc *kms.Client, pk *ecdsa.PublicKey, keyId string, chainID *big.Int) SignerFn {
+	return func(ctx context.Context, address common.Address) (bind.SignerFn, error) {
+		return KMSSignerFn(ctx, svc, pk, keyId, chainID)
+	}
+}
+
+// KMSSignerFn returns a SignerFn that uses a KMS key to sign transactions
+// Heavily taken from https://github.com/welthee/go-ethereum-aws-kms-tx-signer
+func KMSSignerFn(ctx context.Context, svc *kms.Client, pk *ecdsa.PublicKey, keyId string, chainID *big.Int) (bind.SignerFn, error) {
+	if chainID == nil {
+		return nil, errors.New("chainID is required")
+	}
+	if svc == nil {
+		return nil, errors.New("kms client is required")
+	}
+	if pk == nil {
+		return nil, errors.New("public key is required")
+	}
+
+	pubKeyBytes := secp256k1.S256().Marshal(pk.X, pk.Y)
+	keyAddr := crypto.PubkeyToAddress(*pk)
+	signer := types.LatestSignerForChainID(chainID)
+	return func(address common.Address, tx *types.Transaction) (*types.Transaction, error) {
+		if address != keyAddr {
+			return nil, bind.ErrNotAuthorized
+		}
+
+		txHashBytes := signer.Hash(tx).Bytes()
+
+		rBytes, sBytes, err := eigenkms.GetSignature(ctx, svc, keyId, txHashBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		// Adjust S value from signature according to Ethereum standard
+		sBigInt := new(big.Int).SetBytes(sBytes)
+		if sBigInt.Cmp(secp256k1HalfN) > 0 {
+			sBytes = new(big.Int).Sub(secp256k1N, sBigInt).Bytes()
+		}
+
+		signature, err := getEthereumSignature(pubKeyBytes, txHashBytes, rBytes, sBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		return tx.WithSignature(signer, signature)
+	}, nil
+}
+
+func getEthereumSignature(expectedPublicKeyBytes []byte, txHash []byte, r []byte, s []byte) ([]byte, error) {
+	rsSignature := append(adjustSignatureLength(r), adjustSignatureLength(s)...)
+	signature := append(rsSignature, []byte{0}...)
```

Comment:
some info here https://yos.io/2018/11/16/ethereum-signatures/

Reply:
It tries with 0 and then if the recovered pub key doesn't match, it tries again with 1 in L82

Reply:
Yeah but my point was that sometimes it needs to use 27/28 instead of 0/1 right? I've had to manually add 27 to v in some codebase somewhere, keep forgetting why though because I never try to actually understand it.

Reply:
If this is working for you we can just merge with this and then generalize later when someone hits a bug with 0/1. Feel free to resolve this if that works for you.

Reply:
I think 27/28 maps to 0/1 in geth code. It expects V to be either 0 or 1: https://github.com/ethereum/go-ethereum/blob/master/crypto/signature_cgo.go#L52

Reply:
That's their underlying crypto library, but they use a wrapper ontop, can't find it right now though.
Also if we ever use this code to send an ecdsa to an avs contract, its possible they would need 27/28 instead. I remember hitting that issue, but forget details. See https://github.com/Layr-Labs/eigenda/blob/8ec570b8c2b266fad20ea0af14f0f5d84906c39c/core/eth/tx.go#L169 for eg

Reply:
This was driving me insane, but finally got to the bottom of it.

The magic is done by tx.WithSignature() on line 68 above. That is what transforms the 0/1 value into an eip-155 conformant value (27/28 or chain_id*2+35/36), by calling the underlying signer, for eg https://github.com/ethereum/go-ethereum/blob/86a1f0c39494c8f5caddf6bd9fbddd4bdfa944fd/core/types/transaction_signing.go#L331

But basically this means that the KMS signer and the privatekey signer both have very similar structure, and could potentially be refactored to reuse some code (not sure it's worth it, just point out).
That's because both of them receive a value of v being 0/1 from the underlying method, and then use `tx.WithSignature` with an appropriate geth higher-level signer to set the v value properly. This all makes me wonder whether there isn't a simpler way to implement these things (feels so weird that we're implementing a signerFn for abigen that uses a low-level crypto.Sign but then passes the output back to another higher-level embedded signer to fill the V value properly... cc @shrimalmadhur

Reply:
But @ian-shim can you just add a comment above getEthereumSignature mentioning that this function returns a v value of 0/1, because the chainIdSigner `signer := types.LatestSignerForChainID(chainID)` expects this and then turns it into the proper eip-155 value so that the tx is accepted by ethereum?

Then we can merge this.

Reply:
Done